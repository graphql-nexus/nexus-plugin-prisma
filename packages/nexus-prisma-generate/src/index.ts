#!/usr/bin/env node

import { existsSync, mkdirSync, writeFileSync } from 'fs'
import {
  GraphQLEnumType,
  GraphQLField,
  GraphQLInputField,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLSchema,
  isEnumType,
  isInputObjectType,
  isObjectType,
  isScalarType,
  introspectionFromSchema,
} from 'graphql'
import * as meow from 'meow'
import { EOL } from 'os'
import { join } from 'path'
import { ISDL } from 'prisma-datamodel'
import { generateCRUDSchemaFromInternalISDL } from 'prisma-generate-schema'
import {
  findDatamodelAndComputeSchema,
  findRootDirectory,
  getImportPathRelativeToOutput,
} from './config'
import { getFinalType, getTypeName, isList, isRequired } from './graphql'

const cli = meow(
  `
    nexus-prisma-generate prisma-client-dir output

    > Generate the TypeScript types for nexus-prisma

    -----
    
    Inputs should be relative to the root of your project

    --client (required): Path to your prisma-client directory (eg: ./generated/prisma-client/)
    --output (required): Path to directory where you want to output the typings (eg: ./generated/nexus-prisma)
    --js     (optional): Whether to generate the types for Javascript
`,
  {
    flags: {
      client: {
        type: 'string',
      },
      output: {
        type: 'string',
      },
      js: {
        type: 'boolean',
        default: false,
      },
    },
  },
)

main(cli)

function main(cli: meow.Result) {
  const { client: prismaClientDir, output, js: jsMode } = cli.flags

  if (!prismaClientDir) {
    console.log('ERROR: Missing argument --client')
    process.exit(1)
  }

  if (!output) {
    console.log('ERROR: Missing argument --output')
    process.exit(1)
  }

  const rootPath = findRootDirectory()
  const resolvedOutput = output.startsWith('/')
    ? output
    : join(rootPath, output)
  const resolvedPrismaClientDir = prismaClientDir.startsWith('/')
    ? prismaClientDir
    : join(rootPath, prismaClientDir)

  if (!existsSync(resolvedPrismaClientDir)) {
    console.log(
      `ERROR: Cannot find --client path at ${resolvedPrismaClientDir}`,
    )
    process.exit(1)
  }

  try {
    // Create the output directories if needed (mkdir -p)
    mkdirSync(resolvedOutput, { recursive: true })
  } catch (e) {
    if (e.code !== 'EEXIST') throw e
  }

  const { datamodel, databaseType } = findDatamodelAndComputeSchema()

  try {
    const schema = generateCRUDSchemaFromInternalISDL(datamodel, databaseType)
    const renderedDatamodel = renderDatamodel(datamodel, schema)
    const nexusPrismaTypesPath = join(rootPath, output, 'nexus-prisma.ts')
    const nexusPrismaTypes = renderNexusPrismaTypes(
      schema,
      getImportPathRelativeToOutput(
        resolvedPrismaClientDir,
        nexusPrismaTypesPath,
      ),
    )

    writeFileSync(nexusPrismaTypesPath, nexusPrismaTypes)

    if (jsMode) {
      const datamodelPath = join(rootPath, output, 'datamodel-info.js')
      const indexPath = join(rootPath, output, 'index.js')

      writeFileSync(datamodelPath, `module.exports = ${renderedDatamodel}`)
      writeFileSync(indexPath, renderIndexJs())
    } else {
      const datamodelPath = join(rootPath, output, 'datamodel-info.ts')
      const indexPath = join(rootPath, output, 'index.ts')

      writeFileSync(datamodelPath, `export default ${renderedDatamodel}`)
      writeFileSync(indexPath, `export { default } from './datamodel-info'`)
    }

    console.log(`Types generated at ${output}`)
  } catch (e) {
    console.error(e)
  }
}

function renderIndexJs() {
  return `\
const datamodelInfo = require('./datamodel-info')
  
module.exports = datamodelInfo
  `
}

function renderDatamodel(datamodel: ISDL, schema: GraphQLSchema) {
  return `\
{
  uniqueFieldsByModel: {
${datamodel.types
  .map(
    type =>
      `    ${type.name}: [${type.fields
        .filter(field => field.isUnique)
        .map(field => `'${field.name}'`)
        .join(', ')}]`,
  )
  .join(',' + EOL)}
  },
  schema: ${JSON.stringify(introspectionFromSchema(schema), null, 2)}
}
  `
}

function renderNexusPrismaTypes(
  schema: GraphQLSchema,
  prismaClientPath: string,
) {
  const types = Object.values(schema.getTypeMap())
  const objectTypes = types.filter(
    t => isObjectType(t) && !t.name.startsWith('__'),
  ) as GraphQLObjectType[]
  const inputTypes = types.filter(isInputObjectType)
  const enumTypes = types.filter(
    t => isEnumType(t) && !t.name.startsWith('__'),
  ) as GraphQLEnumType[]

  return `\
// GENERATED TYPES FOR NEXUS-PRISMA. /!\\ DO NOT EDIT MANUALLY

import { core } from 'nexus'
import { GraphQLResolveInfo } from 'graphql'
import * as prisma from '${prismaClientPath}'

declare global {
  interface NexusPrismaGen extends NexusPrismaTypes {}
}

export interface NexusPrismaTypes {
  objectTypes: {
    fields: {
${objectTypes
  .map(type => `      ${type.name}: ${getObjectTypeFieldsName(type)}`)
  .join(EOL)}
    }
    fieldsDetails: {
${objectTypes
  .map(type => `      ${type.name}: ${getObjectTypeFieldsDetailsName(type)}`)
  .join(EOL)}
    }
  }
  inputTypes: {
    fields: {
${inputTypes
  .map(
    type =>
      `      ${getInputObjectTypeName(type)}: ${getInputObjectTypeFieldsName(
        type,
      )}`,
  )
  .join(EOL)}
    }
  }
  enumTypes: {
${enumTypes
  .map(type => `    ${type.name}: ${getEnumTypeName(type)},`)
  .join(EOL)}
  }
}

${objectTypes.map(renderObjectType).join(EOL)}

${inputTypes.map(renderInputType).join(EOL)}

${enumTypes.map(renderEnumType).join(EOL)}
  `
}

function renderObjectType(type: GraphQLObjectType) {
  const fields = Object.values(type.getFields())

  return `\
// Types for ${type.name}

${renderFields(type, fields)}

${renderFieldsArgs(type, fields)}

${renderTypeFieldDetails(type, fields)}
`
}

function renderTypeFieldDetails(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
export interface ${getObjectTypeFieldsDetailsName(type)} {
${fields
  .map(
    field => `\
  ${field.name}: {
    type: '${getTypeName(field.type)}'
    args: ${
      field.args.length > 0
        ? `Record<${getTypeFieldArgName(
            type,
            field,
          )}, core.NexusArgDef<string>>`
        : '{}'
    }
    description: string
    list: ${isList(field.type) ? true : undefined}
    nullable: ${!isRequired(field.type)}
    resolve: ${
      isScalarType(getFinalType(field.type))
        ? undefined
        : `(
      root: core.RootValue<"${type.name}">,
      args: ${renderResolverArgs(field)},
      context: core.GetGen<"context">,
      info?: GraphQLResolveInfo
    ) => ${renderResolverReturnType(field)}`
    }
  }`,
  )
  .join(EOL)}
}
  `
}

function renderResolverArgs(field: GraphQLField<any, any>) {
  return `{\
 ${field.args
   .map(
     arg => `${arg.name}${isRequired(arg.type) ? '' : '?'}: ${getTSType(arg)}`,
   )
   .join(', ')} }\
  `
}

function renderFields(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
type ${getObjectTypeFieldsName(type)} =
  | ${getExposableFieldsTypeName(type)}
${fields
  .map(
    f =>
      `  | { name: '${f.name}', args?: ${
        f.args.length > 0
          ? `${getTypeFieldArgName(type, f)}[] | false`
          : '[] | false'
      }, alias?: string  } `,
  )
  .join(EOL)}

type ${getExposableFieldsTypeName(type)} =
${fields.map(f => `  | '${f.name}'`).join(EOL)}
`
}

function renderFieldsArgs(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
${fields
  .filter(field => field.args.length > 0)
  .map(field => renderFieldArg(type, field))
  .join(EOL)}
  `
}

function renderFieldArg(
  type: GraphQLObjectType,
  field: GraphQLField<any, any>,
) {
  return `\
type ${getTypeFieldArgName(type, field)} =
${field.args.map(arg => `  | '${arg.name}'`).join(EOL)}`
}

function getTSType(graphqlType: GraphQLField<any, any> | GraphQLInputField) {
  const graphqlToTypescript: Record<string, string> = {
    String: 'string',
    Boolean: 'boolean',
    ID: 'string',
    Int: 'number',
    Float: 'number',
    DateTime: 'string',
  }
  const finalType = getFinalType(graphqlType.type)

  let returnType = ''

  if (isScalarType(finalType)) {
    returnType = graphqlToTypescript[getTypeName(finalType)]
  } else if (isInputObjectType(finalType)) {
    returnType = getInputObjectTypeName(finalType)
  } else {
    returnType = `prisma.${getTypeName(finalType)}`
  }

  if (isList(graphqlType.type)) {
    returnType += '[]'
  }

  if (!isRequired(graphqlType.type)) {
    returnType += ' | null'
  }

  return returnType
}

function renderResolverReturnType(field: GraphQLField<any, any>) {
  const returnType = getTSType(field)

  return `Promise<${returnType}> | ${returnType}`
}

function renderInputType(input: GraphQLInputObjectType) {
  const fields = Object.values(input.getFields())
  return `\
export interface ${getInputObjectTypeName(input)} {
${fields
  .map(
    field =>
      `  ${field.name}${isRequired(input) ? '' : '?'}: ${getTSType(field)}`,
  )
  .join(EOL)}
}
export type ${getInputObjectTypeFieldsName(input)} =
  | Extract<keyof ${getInputObjectTypeName(input)}, string>
${fields.map(f => `  | { name: '${f.name}', alias?: string  } `).join(EOL)}
  `
}

function renderEnumType(enumType: GraphQLEnumType): string {
  return `\
export type ${getEnumTypeName(enumType)} =
${enumType
  .getValues()
  .map(value => `  | '${value.name}'`)
  .join(EOL)}
  `
}

function getExposableFieldsTypeName(type: GraphQLObjectType) {
  return `${type.name}Fields`
}

function upperFirst(s: string) {
  return s.replace(/^\w/, c => c.toUpperCase())
}

function getTypeFieldArgName(
  type: GraphQLObjectType,
  field: GraphQLField<any, any>,
) {
  return `${type.name}${upperFirst(field.name)}Args`
}

function getObjectTypeFieldsName(type: GraphQLObjectType) {
  return `${type.name}Object`
}

function getInputObjectTypeFieldsName(type: GraphQLInputObjectType) {
  return `${type.name}InputObject`
}

function getObjectTypeFieldsDetailsName(type: GraphQLObjectType) {
  return `${type.name}FieldDetails`
}

function getInputObjectTypeName(type: GraphQLInputObjectType) {
  return `${type.name}`
}

function getEnumTypeName(enumType: GraphQLEnumType): string {
  return `${enumType.name}Values`
}
