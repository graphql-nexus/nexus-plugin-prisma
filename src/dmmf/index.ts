export { get } from './importer'
export { transform } from './transformer'
export { DMMFClass as DMMF } from './DMMFClass'
export { ExternalDMMF as External } from './types'
