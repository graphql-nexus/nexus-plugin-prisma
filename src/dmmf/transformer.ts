import { DMMF, ExternalDMMF } from './types'

export function transform(document: DMMF.Document): ExternalDMMF.Document {
  return {
    datamodel: transformDatamodel(document.datamodel),
    mappings: document.mappings as ExternalDMMF.Mapping[],
    schema: transformSchema(document.schema),
  }
}

function transformDatamodel(datamodel: DMMF.Datamodel): ExternalDMMF.Datamodel {
  return {
    enums: datamodel.enums,
    models: datamodel.models.map(model => ({
      ...model,
      fields: model.fields.map(field => ({
        ...field,
        kind: field.kind === 'object' ? 'relation' : field.kind,
      })),
    })),
  }
}

function transformSchema(schema: DMMF.Schema): ExternalDMMF.Schema {
  return {
    enums: schema.enums,
    inputTypes: schema.inputTypes.map(transformInputType),
    outputTypes: schema.outputTypes.map(o => ({
      ...o,
      fields: o.fields.map(f => ({
        ...f,
        args: f.args.map(transformArg),
        outputType: {
          ...f.outputType,
          type: getReturnTypeName(f.outputType.type),
        },
      })),
    })),
  }
}

function transformArg(arg: DMMF.SchemaArg): ExternalDMMF.SchemaArg {
  // FIXME: *Enum*Filter are currently empty
  let inputType = arg.inputType.some(a => a.kind === 'enum')
    ? arg.inputType[0]
    : arg.inputType.find(a => a.kind === 'object')!

  if (!inputType) {
    inputType = arg.inputType[0]
  }

  return {
    name: arg.name,
    inputType: {
      ...inputType,
      type: getReturnTypeName(inputType.type),
    },
    isRelationFilter: undefined,
  }
}

function transformInputType(inputType: DMMF.InputType): ExternalDMMF.InputType {
  return {
    ...inputType,
    fields: inputType.fields.map(transformArg),
  }
}

function getReturnTypeName(type: any) {
  if (typeof type === 'string') {
    return type
  }

  return type.name
}
