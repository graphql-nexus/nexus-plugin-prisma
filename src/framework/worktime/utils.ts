import * as Prisma from '@prisma/sdk'
import chalk from 'chalk'
import { WorktimeLens } from 'nexus/plugin'
import stripAnsi from 'strip-ansi'
import * as Scaffolders from './scaffolders'

/**
 * Pinned query-engine version. Calculated at build time and based on `@prisma/cli` version
 */
export const PRISMA_QUERY_ENGINE_VERSION: string = require('@prisma/cli/package.json').prisma.version

/**
 * Get the declared generator blocks in the user's PSL file.
 *
 * If Prisma SDK throws a no-models-found error then it is caught, a model is
 * scaffolded, and this function is re-run.
 */
export async function getGenerators(nexus: WorktimeLens, schemaPath: string): Promise<Prisma.Generator[]> {
  return Prisma.getGenerators({
    schemaPath,
    printDownloadProgress: false,
    version: PRISMA_QUERY_ENGINE_VERSION,
  })
    .catch(async (e: Error) => {
      if (isNoModelsDefinedError(e)) {
        await Scaffolders.exampleModelBlock(schemaPath)
        nexus.log.warn(
          `An example model has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        return getGenerators(nexus, schemaPath)
      }
      return Promise.reject(e)
    })
    .then(async (generators) => {
      const hasClientGenerator = generators.find((g) => g.options?.generator.provider === 'prisma-client-js')
      if (!hasClientGenerator) {
        await Scaffolders.prismaClientGeneratorBlock(schemaPath)
        nexus.log.warn(
          `A Prisma Client generator block has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        // TODO: Generate it programmatically instead for performance reason
        return getGenerators(nexus, schemaPath)
      } else {
        return generators
      }
    })
}

/**
 * Check if the thrown error is of NoModelsDefined type.
 */
function isNoModelsDefinedError(e: Error): boolean {
  return Boolean(stripAnsi(e.message).match(/.*You don't have any models defined.*/))
}
