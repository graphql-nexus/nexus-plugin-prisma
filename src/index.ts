import * as Builder from './builder'

export type NexusPrismaParams = Builder.Options
export const nexusPrismaPlugin = Builder.build
