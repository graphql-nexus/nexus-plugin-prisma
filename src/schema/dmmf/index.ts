export * from './utils'
export * from './transformer'
export * from './DmmfDocument'
export * from './DmmfTypes'
