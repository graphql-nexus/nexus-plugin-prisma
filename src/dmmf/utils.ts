import * as Photon from '@prisma/photon'
import { transform } from './transformer'
import { DMMFClass } from './DMMFClass'

export function fromPhotonDMMF(photonDMMF: Photon.DMMF.Document): DMMFClass {
  return new DMMFClass(transform(photonDMMF))
}

export const get = (photonClientPackagePath: string): DMMFClass => {
  let photonClientPackage
  try {
    photonClientPackage = require(photonClientPackagePath)
  } catch (error) {
    throw new Error(
      `Could not find photon package at ${photonClientPackagePath}. Check that you have configured your Photon generator block in schema.prisma correctly and run prisma generate.`,
    )
  }
  return fromPhotonDMMF(photonClientPackage.dmmf)
}
