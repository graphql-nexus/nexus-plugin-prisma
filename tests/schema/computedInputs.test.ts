import * as Nexus from 'nexus'
import { addComputedInputs } from '../../src/dmmf/transformer'
import { GlobalComputedInputs } from '../../src/utils'
import { generateSchemaAndTypes, getDmmf } from '../__utils'

// TODO: Split local and global computedInputs into their own suites

const resolverTestData = {
  datamodel: `
  model User {
    id  Int @id @default(autoincrement())
    name String
    createdWithBrowser String
  }
`,
  query: Nexus.queryType({
    definition(t: any) {
      t.crud.user()
    },
  }),
  mutation: Nexus.mutationType({
    definition(t: any) {
      t.crud.createOneUser({
        computedInputs: {
          createdWithBrowser: ({ ctx }) => ctx.browser,
        } as GlobalComputedInputs,
      })
    },
  }),
  user: Nexus.objectType({
    name: 'User',
    definition: (t: any) => {
      t.model.id()
      t.model.name()
      t.model.createdWithBrowser()
    },
  }),
}

const globalTestData = {
  datamodel: `
  model User {
    id  Int @id @default(autoincrement())
    name String
    nested Nested[]
    createdWithBrowser String
  }

  model Nested {
    id Int @id @default(autoincrement())
    name String
    createdWithBrowser String
  }
`,
  query: Nexus.queryType({
    definition(t: any) {
      t.crud.user()
    },
  }),
  mutation: Nexus.mutationType({
    definition(t: any) {
      t.crud.createOneUser()
      t.crud.createOneNested()
    },
  }),
  user: Nexus.objectType({
    name: 'User',
    definition: (t: any) => {
      t.model.id()
      t.model.name()
      t.model.nested()
      t.model.createdWithBrowser()
    },
  }),
  nested: Nexus.objectType({
    name: 'Nested',
    definition: (t: any) => {
      t.model.id()
      t.model.createdWithBrowser()
      t.model.name()
    },
  }),
}

it.only('removes resolver-level computedInputs from the corresponding input type', async () => {
  const { datamodel, ...resolvers } = resolverTestData
  const result = await generateSchemaAndTypes(
    datamodel,
    Object.values(resolvers),
  )

  expect({
    schema: result.schemaString,
    typegen: result.typegen,
  }).toMatchSnapshot('locallyComputedInputs')
})

it('infers the value of resolver-level computedInputs at runtime', async () => {
  const { datamodel } = resolverTestData
  const dmmf = await getDmmf(datamodel)
  expect(
    addComputedInputs({
      params: {
        info: {} as any,
        args: { data: { name: 'New User' } },
        ctx: { browser: 'firefox' },
      },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      locallyComputedInputs: {
        createdWithBrowser: ({ ctx }) => ctx.browser,
      },
    }),
  ).toStrictEqual({ data: { name: 'New User', createdWithBrowser: 'firefox' } })
})

it('removes global computedInputs from all input types', async () => {
  const { datamodel, ...resolvers } = globalTestData
  const result = await generateSchemaAndTypes(
    datamodel,
    Object.values(resolvers),
    {
      globallyComputedInputs: { createdWithBrowser: ({ ctx }) => ctx.browser },
    },
  )

  expect({
    schema: result.schemaString,
    typegen: result.typegen,
  }).toMatchSnapshot('globallyComputedInputs')
})

it('infers the value of global computedInputs at runtime', async () => {
  const { datamodel } = globalTestData
  const dmmf = await getDmmf(datamodel, {
    globallyComputedInputs: { createdWithBrowser: ({ ctx }) => ctx.browser },
  })
  expect(
    addComputedInputs({
      params: {
        info: {} as any,
        args: {
          data: {
            name: 'New User',
            nested: { create: { name: 'Nested Name' } },
          },
        },
        ctx: { browser: 'firefox' },
      },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      locallyComputedInputs: {},
    }),
  ).toStrictEqual({
    data: {
      name: 'New User',
      createdWithBrowser: 'firefox',
      nested: {
        create: { createdWithBrowser: 'firefox', name: 'Nested Name' },
      },
    },
  })
})

it('handles arrays when recursing for computedInputs', async () => {
  const { datamodel } = globalTestData
  const dmmf = await getDmmf(datamodel, {
    globallyComputedInputs: { createdWithBrowser: ({ ctx }) => ctx.browser },
  })

  expect(
    addComputedInputs({
      params: {
        info: {} as any,
        args: {
          data: {
            name: 'New User',
            nested: {
              create: [{ name: 'Nested Name' }, { name: 'Nested Name' }],
            },
          },
        },
        ctx: { browser: 'firefox' },
      },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      locallyComputedInputs: {},
    }),
  ).toStrictEqual({
    data: {
      name: 'New User',
      createdWithBrowser: 'firefox',
      nested: {
        create: [
          { createdWithBrowser: 'firefox', name: 'Nested Name' },
          { createdWithBrowser: 'firefox', name: 'Nested Name' },
        ],
      },
    },
  })
})

it('can combine resolver-level (shallow) and global (deep) computed inputs', async () => {
  const { datamodel } = globalTestData
  const dmmf = await getDmmf(datamodel, {
    // These are applied globally
    globallyComputedInputs: { createdWithBrowser: ({ ctx }) => ctx.browser },
  })
  expect(
    addComputedInputs({
      params: {
        info: {} as any,
        // name should be required when creating Nested since the computedInput providing
        // it is specific to UserCreateInput and therefore shallow
        args: { data: { nested: { create: { name: 'Nested Name' } } } },
        ctx: { browser: 'firefox', name: 'autopopulated' },
      },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      // These are applied only to UserCreateInput
      locallyComputedInputs: { name: ({ ctx }) => ctx.name },
    }),
  ).toStrictEqual({
    data: {
      name: 'autopopulated',
      createdWithBrowser: 'firefox',
      nested: {
        create: { createdWithBrowser: 'firefox', name: 'Nested Name' },
      },
    },
  })
})

it('can use a combination of args, context and info to compute values', async () => {
  const { datamodel } = globalTestData
  // Nonsense example, but ensures args, ctx and info values are being passed everywhere :)
  const dmmf = await getDmmf(datamodel, {
    globallyComputedInputs: {
      createdWithBrowser: ({ args, ctx, info }) =>
        `${ctx.browser.slice(1, 2)} ${info} ${
          (args.data as any).nested.create.name
        }`,
    },
  })
  expect(
    addComputedInputs({
      params: {
        // Normally this would be GraphQLResolveInfo but using a string for simplicity
        info: 'Yam' as any,
        args: { data: { nested: { create: { name: 'Sam' } } } },
        ctx: { browser: 'firefox' },
      },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      locallyComputedInputs: {
        name: ({ args, ctx, info }) =>
          `${args.data.nested.create.name} ${ctx.browser.slice(1, 2)} ${info}`,
      },
    }),
  ).toStrictEqual({
    data: {
      name: 'Sam i Yam',
      createdWithBrowser: 'i Yam Sam',
      nested: {
        create: { createdWithBrowser: 'i Yam Sam', name: 'Sam' },
      },
    },
  })
})
