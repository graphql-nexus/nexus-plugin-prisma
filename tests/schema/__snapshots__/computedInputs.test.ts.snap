// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removes global computedInputs from all input types: globallyComputedInputs 1`] = `
Object {
  "schema": "type Query {
  user(where: UserWhereUniqueInput!): User
}

type Mutation {
  createOneUser(data: UserCreateInput!): User!
  createOneNested(data: NestedCreateInput!): Nested!
}

type User {
  id: Int!
  name: String!
  nested(first: Int, last: Int, before: NestedWhereUniqueInput, after: NestedWhereUniqueInput): [Nested!]!
  createdWithBrowser: String!
}

type Nested {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input UserWhereUniqueInput {
  id: Int
}

input UserCreateInput {
  name: String!
  nested: NestedCreateNestedManyWithoutUserInput
}

input NestedCreateInput {
  name: String!
  User: UserCreateNestedOneWithoutNestedInput
}

input NestedWhereUniqueInput {
  id: Int
}

input NestedCreateNestedManyWithoutUserInput {
  create: [NestedCreateWithoutUserInput!]
  connectOrCreate: [NestedCreateOrConnectWithoutUserInput!]
  connect: [NestedWhereUniqueInput!]
}

input UserCreateNestedOneWithoutNestedInput {
  create: UserCreateWithoutNestedInput
  connectOrCreate: UserCreateOrConnectWithoutnestedInput
  connect: UserWhereUniqueInput
}

input NestedCreateWithoutUserInput {
  name: String!
}

input NestedCreateOrConnectWithoutUserInput {
  where: NestedWhereUniqueInput!
  create: NestedCreateWithoutUserInput!
}

input UserCreateWithoutNestedInput {
  name: String!
}

input UserCreateOrConnectWithoutnestedInput {
  where: UserWhereUniqueInput!
  create: UserCreateWithoutNestedInput!
}
",
  "typegen": "import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
    first?: boolean
    last?: boolean
    before?: boolean
    after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  User: Prisma.User
  Nested: Prisma.Nested
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    users: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'name' | 'nested'
      ordering: 'id' | 'name'
    }
    nesteds: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'name' | 'User' | 'userId'
      ordering: 'id' | 'name' | 'User' | 'userId'
    }
  },
  User: {
    nested: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'name' | 'User' | 'userId'
      ordering: 'id' | 'name' | 'User' | 'userId'
    }
  }
  Nested: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    user: 'User'
    users: 'User'
    nested: 'Nested'
    nesteds: 'Nested'
  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'AffectedRowsOutput'
    deleteOneUser: 'User'
    deleteManyUser: 'AffectedRowsOutput'
    upsertOneUser: 'User'
    createOneNested: 'Nested'
    updateOneNested: 'Nested'
    updateManyNested: 'AffectedRowsOutput'
    deleteOneNested: 'Nested'
    deleteManyNested: 'AffectedRowsOutput'
    upsertOneNested: 'Nested'
  },
  User: {
    id: 'Int'
    name: 'String'
    nested: 'Nested'
    createdWithBrowser: 'String'
  }
  Nested: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    User: 'User'
    userId: 'Int'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  User: Typegen.NexusPrismaFields<'User'>
  Nested: Typegen.NexusPrismaFields<'Nested'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  ",
}
`;

exports[`removes resolver-level computedInputs from the corresponding input type: locallyComputedInputs 1`] = `
Object {
  "schema": "type Query {
  user(where: UserWhereUniqueInput!): User
}

type Mutation {
  createOneUser(data: UserCreateInput!): User!
}

type User {
  id: Int!
  name: String!
  createdWithBrowser: String!
}

input UserWhereUniqueInput {
  id: Int
}

input UserCreateInput {
  name: String!
}
",
  "typegen": "import * as Typegen from 'nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
    first?: boolean
    last?: boolean
    before?: boolean
    after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  User: Prisma.User
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    users: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'name' | 'createdWithBrowser'
      ordering: 'id' | 'name' | 'createdWithBrowser'
    }
  },
  User: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    user: 'User'
    users: 'User'
  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'AffectedRowsOutput'
    deleteOneUser: 'User'
    deleteManyUser: 'AffectedRowsOutput'
    upsertOneUser: 'User'
  },
  User: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  User: Typegen.NexusPrismaFields<'User'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  ",
}
`;
